javalib/src/main/scala/java/lang/Integer.scala

```diff
@@ -82,46 +115,88 @@ object Integer {
   @noinline def parseUnsignedInt(s: String, radix: scala.Int): scala.Int =
     parseIntImpl(s, radix, signed = false)
 
+  private def parseIntFail(s: String): Nothing =
+    throw new NumberFormatException(s"""For input string: "$s"""")
+
   @inline
   private def parseIntImpl(s: String, radix: scala.Int,
       signed: scala.Boolean): scala.Int = {
 
-    def fail(): Nothing =
-      throw new NumberFormatException(s"""For input string: "$s"""")
+    def fail(): Nothing = parseIntFail(s)
 
-    val len = if (s == null) 0 else s.length
-
-    if (len == 0 || radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
+    if (s == null || s == "")
       fail()
 
     val firstChar = s.charAt(0)
     val negative = signed && firstChar == '-'
-
-    val maxAbsValue: scala.Double = {
-      if (!signed) 0xffffffffL.toDouble
-      else if (negative) 0x80000000L.toDouble
-      else 0x7fffffffL.toDouble
+    val start = if (negative || firstChar == '+') 1 else 0
+
+    val unsignedResult = parseIntInternal(s, radix, start)
+
+    if (signed) {
+      if (negative) {
+        val result = -unsignedResult
+        if (result > 0)
+          fail()
+        result
+      } else {
+        if (unsignedResult < 0)
+          fail()
+        unsignedResult
+      }
+    } else {
+      unsignedResult
     }
+  }
+
+  @noinline
+  private def parseIntInternal(s: String, radix: Int, start: Int): Int = {
+    def fail(): Nothing = parseIntFail(s)
 
-    var i = if (negative || firstChar == '+') 1 else 0
+    val len = s.length()
 
-    // We need at least one digit
-    if (i >= s.length)
+    // We need at least one digit, even if it is a 0 digit; also check the radix
+    if (start == len || radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
```

Comment:
```suggestion
    if (start >= len || radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
```

For clarity / safety?